File: BackEnd/src/data/mappers/reminder_mapper.ts
/**
 * @file reminder_mapper.ts
 * @description Implementación del mapper de recordatorios de riego.
 * Convierte entre ReminderDocument (MongoDB), Reminder (dominio) y ReminderResponseDTO (presentación).
 * @module Reminders
 * @layer Data
 *
 * @implements {IReminderMapper}
 * @injectable
 */

import { injectable } from 'inversify';
import { ObjectId } from 'mongodb';
import type { IReminderMapper } from '../IMappers/IReminderMapper.js';
import { Reminder, ReminderType } from '../../domain/entities/Reminder.js';
import { ReminderDocument } from '../datasources/mongodb/models/ReminderModel.js';
import type { ReminderResponseDTO } from '../../domain/dtos/reminders/reminder-response.dto.js';

/**
 * Mapper de recordatorios de riego.
 *
 * @implements {IReminderMapper}
 * @injectable
 */
@injectable()
export class ReminderMapper implements IReminderMapper {

  /**
   * Convierte un documento MongoDB a entidad Reminder.
   *
   * @param doc — Documento de la colección 'reminders'.
   * @returns Entidad Reminder.
   */
  toEntity(doc: ReminderDocument): Reminder {
    return new Reminder({
      id:            doc._id.toHexString(),
      plantId:       doc.plantId.toHexString(),
      userId:        doc.userId.toHexString(),
      type:          doc.type as ReminderType,
      scheduledDate: doc.scheduledDate,
      message:       doc.message,
      isCompleted:   doc.isCompleted,
      suspended:     doc.suspended,
      attempts:      doc.attempts,
      createdAt:     doc.createdAt,
    });
  }

  /**
   * Convierte una entidad Reminder a documento MongoDB.
   *
   * @param entity — Entidad Reminder.
   * @returns ReminderDocument para insertar/actualizar.
   */
  toDocument(entity: Reminder): ReminderDocument {
    return {
      _id:           new ObjectId(entity.id),
      plantId:       new ObjectId(entity.plantId),
      userId:        new ObjectId(entity.userId),
      type:          entity.type,
      scheduledDate: entity.scheduledDate,
      message:       entity.message,
      isCompleted:   entity.isCompleted,
      suspended:     entity.suspended,
      attempts:      entity.attempts,
      createdAt:     entity.createdAt,
    };
  }

  /**
   * Convierte una entidad Reminder al DTO de respuesta HTTP.
   *
   * @param entity — Entidad Reminder.
   * @returns ReminderResponseDTO serializable.
   */
  toResponseDTO(entity: Reminder): ReminderResponseDTO {
    return {
      id:            entity.id,
      plantId:       entity.plantId,
      userId:        entity.userId,
      type:          entity.type,
      scheduledDate: entity.scheduledDate.toISOString(),
      message:       entity.message,
      isCompleted:   entity.isCompleted,
      suspended:     entity.suspended,
      attempts:      entity.attempts,
      createdAt:     entity.createdAt.toISOString(),
      isPending:     entity.isPending,
    };
  }
}

File: FrontEnd/lib/data/mappers/reminder_mapper.dart
/// @file reminder_mapper.dart
/// @description Implementación del mapper de recordatorios.
/// Convierte ReminderModel ↔ Reminder normalizando tipos y fechas.
/// @module Reminders
/// @layer Data
library;

import '../../core/utils/date_utils.dart';
import '../../domain/entities/reminder.dart';
import '../i_mappers/i_reminder_mapper.dart';
import '../models/reminder_model.dart';

// ═══════════════════════════════════════════════════════════════════════════════
// REMINDER MAPPER
// ═══════════════════════════════════════════════════════════════════════════════

/// Implementación de [IReminderMapper].
///
/// [implements] IReminderMapper
/// [injectable] registrar en container.dart como singleton.
class ReminderMapper implements IReminderMapper {

  // ─── ReminderModel → Reminder ─────────────────────────────────────────────────

  @override
  Reminder toEntity(ReminderModel model) {
    return Reminder(
      id:            model.id,
      plantId:       model.plantId,
      userId:        model.userId,
      type:          model.type,
      scheduledDate: PlantDateUtils.parseUtc(model.scheduledDate) ?? DateTime.now().toUtc(),
      message:       model.message,
      isCompleted:   model.isCompleted,
      suspended:     model.suspended,
      attempts:      model.attempts,
      createdAt:     PlantDateUtils.parseUtc(model.createdAt) ?? DateTime.now().toUtc(),
    );
  }

  // ─── Reminder → ReminderModel ─────────────────────────────────────────────────

  @override
  ReminderModel toModel(Reminder entity) {
    return ReminderModel(
      id:            entity.id,
      plantId:       entity.plantId,
      userId:        entity.userId,
      type:          entity.type,
      scheduledDate: PlantDateUtils.toIso8601(entity.scheduledDate),
      message:       entity.message,
      isCompleted:   entity.isCompleted,
      suspended:     entity.suspended,
      attempts:      entity.attempts,
      createdAt:     PlantDateUtils.toIso8601(entity.createdAt),
    );
  }
}
